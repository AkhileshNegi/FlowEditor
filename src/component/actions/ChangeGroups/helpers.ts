import { Group, ChangeGroups } from '../../../flowTypes';
import { SearchResult } from '../../../store';

export const mapGroupsToSearchResults = (groups: Group[]) =>
    groups.map(({ name, uuid }) => ({ name, id: uuid }));

export const mapSearchResultsToGroups = (searchResults: SearchResult[]) =>
    searchResults.map(result => ({
        uuid: result.id,
        name: result.name
    }));

export const getGroups = (action: ChangeGroups): SearchResult[] => {
    if (action.groups == null) {
        return [];
    }

    if (action.groups.length && action.type !== 'add_contact_groups') {
        return mapGroupsToSearchResults(action.groups);
    }

    return [];
};
