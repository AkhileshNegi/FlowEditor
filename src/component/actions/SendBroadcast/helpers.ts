import { BroadcastMsg, Contact, Group } from '../../../flowTypes';
import { Asset, AssetType } from '../../../services/AssetService';

export const getRecipients = (action: BroadcastMsg): Asset[] => {
    const selected = action.groups.map((group: Group) => {
        return { id: group.uuid, name: group.name, type: AssetType.Group };
    });

    return selected.concat(
        action.contacts.map((contact: Contact) => {
            return { id: contact.uuid, name: contact.name, type: AssetType.Contact };
        })
    );
};
