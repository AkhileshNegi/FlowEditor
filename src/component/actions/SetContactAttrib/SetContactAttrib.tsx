import * as React from 'react';
import { SetContactField, SetContactProperty } from '../../../flowTypes';

const SetContactAttribComp: React.SFC<SetContactField & SetContactProperty> = ({
    field_name,
    property_name,
    value
}) => {
    const fieldNameMarkup = <span className="emph">{field_name || property_name}</span>;
    const valueMarkup = <span className="emph">{value}</span>;
    if (value) {
        return (
            <div>
                Update {fieldNameMarkup} to {valueMarkup}
            </div>
        );
    }
    return <div>Clear value for {fieldNameMarkup}</div>;
};

export default SetContactAttribComp;
