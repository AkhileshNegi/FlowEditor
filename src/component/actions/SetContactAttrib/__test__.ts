import {
    ContactProperties,
    SetContactField,
    SetContactProperty,
    SendMsg
} from '../../../flowTypes';

export const setContactProperty: SetContactProperty = {
    type: 'set_contact_property',
    uuid: '4fac7935-d13b-4b36-bf15-98075dca822a',
    property: ContactProperties.Name.toLowerCase(),
    value: 'Jane Doe'
};

export const setContactField: SetContactField = {
    type: 'set_contact_field',
    uuid: '445fc64c-2a18-47cc-89d0-15172826bfcc',
    field: {
        key: 'age',
        name: 'Age'
    },
    value: '25'
};
