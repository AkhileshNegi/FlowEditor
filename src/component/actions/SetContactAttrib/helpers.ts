import { Types } from '../../../config/typeConfigs';
import { Action, SetContactField, SetContactName, SetContactProperty } from '../../../flowTypes';
import { Asset, AssetType } from '../../../services/AssetService';
import { snakify, titleCase } from '../../../utils';

export type ContactProperty = Types.set_contact_name;

export const newFieldAction = ({
    uuid,
    value,
    name
}: {
    uuid: string;
    value: string;
    name: string;
}): SetContactField => ({
    type: Types.set_contact_field,
    field: {
        key: snakify(name),
        name: titleCase(name)
    },
    uuid,
    value
});

export const newPropertyAction = ({
    uuid,
    value,
    type
}: {
    uuid: string;
    value: string;
    type: AssetType;
}): SetContactProperty => {
    const action: Partial<Action> = {
        uuid
    };
    switch (type) {
        case AssetType.Name:
            return {
                ...action,
                type: Types.set_contact_name,
                name: value
            } as SetContactName;
    }
};

export const fieldToAsset = ({ field: { key, name } }: SetContactField): Asset => ({
    id: key,
    name,
    type: AssetType.Field
});

export const propertyToAsset = (property: SetContactProperty): Asset => {
    switch (property.type) {
        case Types.set_contact_name:
            return {
                type: AssetType.Name,
                name: titleCase(AssetType.Name),
                id: AssetType.Name
            };
    }
};
