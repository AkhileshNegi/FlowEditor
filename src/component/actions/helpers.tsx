import * as React from 'react';

import { Asset, AssetType } from '../../services/AssetService';

const styles = require('../shared.scss');

export const renderAssetList = (assets: Asset[], max: number): JSX.Element[] => {
    return assets.reduce((elements, asset, idx) => {
        if (idx <= max - 1) {
            elements.push(renderAsset(asset));
        } else if (idx > max - 1 && elements.length === max) {
            elements.push(<div key="ellipses">...</div>);
        }
        return elements;
    }, []);
};

export const renderAsset = (asset: Asset) => {
    switch (asset.type) {
        case AssetType.Group:
            return (
                <div className={styles.nodeAsset} key={asset.id}>
                    <span className={`${styles.nodeGroup} icn-group`} />
                    {asset.name}
                </div>
            );
        case AssetType.Label:
            return (
                <div className={styles.nodeAsset} key={asset.id}>
                    <span className={`${styles.nodeLabel} icn-label`} />
                    {asset.name}
                </div>
            );
    }

    return (
        <div className={styles.nodeAsset} key={asset.id}>
            {asset.name}
        </div>
    );
};
