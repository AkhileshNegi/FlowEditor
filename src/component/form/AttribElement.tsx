import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { getTypeConfig, Type, Types } from '../../config/typeConfigs';
import { CreateOptions, ResultType } from '../../flowTypes';
import { Asset, Assets, AssetType } from '../../services/AssetService';
import {
    AppState,
    DispatchWithState,
    HandleTypeConfigChange,
    handleTypeConfigChange
} from '../../store';
import {
    AssetEntry,
    SetContactFieldFormState,
    SetContactLanguageFormState,
    SetContactNameFormState
} from '../../store/nodeEditor';
import {
    composeCreateNewOption,
    getSelectClassForEntry,
    isOptionUnique,
    isValidNewOption,
    snakify
} from '../../utils';
import SelectSearch from '../SelectSearch/SelectSearch';
import FormElement, { FormElementProps } from './FormElement';

export interface AttribElementPassedProps extends FormElementProps {
    assets: Assets;
    onChange(selected: Asset): void;

    add?: boolean;
    placeholder?: string;
    searchPromptText?: string;
    helpText?: string;
}

export interface AttribElementStoreProps {
    attribute: AssetEntry;
    typeConfig: Type;
    handleTypeConfigChange: HandleTypeConfigChange;
}

export type AttribElementProps = AttribElementPassedProps & AttribElementStoreProps;

export const PLACEHOLDER = 'Enter the name of an existing attribute or create a new one';
export const NOT_FOUND = 'Invalid attribute';
export const CREATE_PROMPT = 'New attribute: ';

export const createNewOption = composeCreateNewOption({
    idCb: label => snakify(label),
    type: AssetType.Field
});

export class AttribElement extends React.Component<AttribElementProps> {
    public static defaultProps = {
        placeholder: PLACEHOLDER,
        searchPromptText: NOT_FOUND
    };

    constructor(props: any) {
        super(props);

        this.onChange = this.onChange.bind(this);
    }

    private onChange(selected: Asset[]): void {
        const [attribute] = selected;
        let nextConfig;

        switch (attribute.type) {
            case AssetType.Name:
                nextConfig = getTypeConfig(Types.set_contact_name);
                break;
            case AssetType.Field:
                nextConfig = getTypeConfig(Types.set_contact_field);
                break;
            case AssetType.Language:
                nextConfig = getTypeConfig(Types.set_contact_language);
                break;
        }

        this.props.handleTypeConfigChange(nextConfig, null);

        if (this.props.onChange) {
            this.props.onChange(attribute);
        }
    }

    public render(): JSX.Element {
        const createOptions: CreateOptions = {};

        if (this.props.add) {
            createOptions.isValidNewOption = isValidNewOption;
            createOptions.isOptionUnique = isOptionUnique;
            createOptions.createNewOption = createNewOption;
            createOptions.createPrompt = CREATE_PROMPT;
        }

        return (
            <FormElement
                showLabel={this.props.showLabel}
                name={this.props.name}
                helpText={this.props.helpText}
                entry={this.props.attribute}
                // attribError={this.state.errors.length > 0 }
            >
                <SelectSearch
                    __className={getSelectClassForEntry(this.props.entry)}
                    onChange={this.onChange}
                    name={this.props.name}
                    resultType={ResultType.field}
                    multi={false}
                    assets={this.props.assets}
                    initial={[this.props.attribute.value]}
                    closeOnSelect={true}
                    searchPromptText={this.props.searchPromptText}
                    placeholder={this.props.placeholder}
                    {...createOptions}
                />
            </FormElement>
        );
    }
}

/* istanbul ignore next */
const mapStateToProps = ({ nodeEditor: { form } }: AppState) => ({
    attribute:
        (form as SetContactFieldFormState).field ||
        (form as SetContactNameFormState).name ||
        (form as SetContactLanguageFormState).language
});

/* istanbul ignore next */
const mapDispatchToProps = (dispatch: DispatchWithState) =>
    bindActionCreators(
        {
            handleTypeConfigChange
        },
        dispatch
    );

const ConnectedAttribElement = connect<
    {},
    {
        handleTypeConfigChange: HandleTypeConfigChange;
    },
    AttribElementPassedProps
>(mapStateToProps, mapDispatchToProps, null, {
    withRef: true
})(AttribElement);

export default ConnectedAttribElement;
