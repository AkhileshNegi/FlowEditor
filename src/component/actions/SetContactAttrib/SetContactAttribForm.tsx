import * as React from 'react';
import { ConfigProviderContext } from '../../../config';
import { Types } from '../../../config/typeConfigs';
import { SetContactAttribute, SetContactField, SetContactProperty } from '../../../flowTypes';
import ConnectedAttribElement from '../../form/AttribElement';
import ConnectedTextInputElement from '../../form/TextInputElement';
import { newFieldAction, newPropertyAction, fieldToAsset, propertyToAsset } from './helpers';
import { fakePropType } from '../../../config/ConfigProvider';
import AssetService, { Asset, AssetType } from '../../../services/AssetService';

export interface SetContactAttribFormProps {
    action: SetContactAttribute;
    onBindWidget: (ref: any) => void;
    updateAction: (action: SetContactAttribute) => void;
}

export const ATTRIB_HELP_TEXT =
    'Select an existing attribute to update or type any name to create a new one';
export const TEXT_INPUT_HELP_TEXT =
    'The value to store can be any text you like. You can also reference other values that have been collected up to this point by typing @run.results or @webhook.json.';

export default class SetContactAttribForm extends React.Component<SetContactAttribFormProps> {
    public static contextTypes = {
        endpoints: fakePropType,
        assetService: fakePropType
    };

    constructor(props: SetContactAttribFormProps, context: ConfigProviderContext) {
        super(props);
        this.onValid = this.onValid.bind(this);
    }

    public onValid(widgets: { [name: string]: any }): void {
        const { wrappedInstance: { state: { attribute } } } = widgets.Attribute;
        const { wrappedInstance: { state: { value } } } = widgets.Value;

        if (attribute.type === AssetType.Field) {
            // include our contact field in our local storage
            const assetService: AssetService = this.context.assetService;
            assetService.getFieldAssets().add(attribute);
            this.props.updateAction(newFieldAction(this.props.action.uuid, value, attribute.name));
        } else {
            this.props.updateAction(
                newPropertyAction(this.props.action.uuid, value, attribute.name)
            );
        }
    }

    private getInitial(): Asset {
        if (this.props.action.type === Types.set_contact_field) {
            return fieldToAsset(this.props.action as SetContactField);
        } else {
            return propertyToAsset(this.props.action as SetContactProperty);
        }
    }

    public render(): JSX.Element {
        const initial = this.getInitial();
        return (
            <>
                <ConnectedAttribElement
                    ref={this.props.onBindWidget}
                    name="Attribute"
                    showLabel={true}
                    assets={this.context.assetService.getFieldAssets()}
                    helpText={ATTRIB_HELP_TEXT}
                    initial={initial}
                    add={true}
                    required={true}
                />
                <ConnectedTextInputElement
                    ref={this.props.onBindWidget}
                    name="Value"
                    showLabel={true}
                    value={this.props.action.value}
                    helpText={TEXT_INPUT_HELP_TEXT}
                    autocomplete={true}
                />
            </>
        );
    }
}
