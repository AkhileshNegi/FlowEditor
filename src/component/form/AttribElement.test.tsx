import { getTypeConfig } from '../../config';
import { Types } from '../../config/typeConfigs';
import { Asset, AssetType } from '../../services/AssetService';
import { composeComponentTestUtils, configProviderContext, setMock } from '../../testUtils';
import {
    createSetContactFieldAction,
    createSetContactNameAction
} from '../../testUtils/assetCreators';
import { isOptionUnique, isValidNewOption } from '../../utils';
import { fieldToAsset, propertyToAsset } from '../actions/SetContactAttrib/helpers';
import { AttribElement, AttribElementProps, CREATE_PROMPT, createNewOption } from './AttribElement';

const attribute: Asset = {
    id: 'name',
    name: 'Name',
    type: AssetType.Name
};

const baseProps: AttribElementProps = {
    name: 'Attribute',
    attribute: { value: attribute },
    assets: configProviderContext.assetService.getFieldAssets(),
    updateTypeConfig: jest.fn(),
    onChange: jest.fn()
};

const { setup, spyOn } = composeComponentTestUtils(AttribElement, baseProps);

describe(AttribElement.name, () => {
    describe('helpers', () => {
        describe('createNewOption', () => {
            it('should return a new SearchResult', () => {
                const newOption = { label: 'Home Phone', labelKey: 'name', valueKey: 'id' };

                expect(createNewOption(newOption)).toEqual({
                    id: 'home_phone',
                    name: newOption.label,
                    type: AssetType.Field,
                    isNew: true
                });
            });
        });
    });

    describe('render', () => {
        it('should pass createOptions to SelectSearch', () => {
            const { wrapper } = setup(true, { add: { $set: true } });

            expect(wrapper.find('SelectSearch').props()).toEqual(
                expect.objectContaining({
                    isValidNewOption,
                    isOptionUnique,
                    createNewOption,
                    createPrompt: CREATE_PROMPT
                })
            );
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('instance methods', () => {
        describe('onChange', () => {
            it('should update type config if passed an contact name asset', () => {
                const { wrapper, instance, props } = setup(true, { updateTypeConfig: setMock() });

                instance.onChange([props.attribute.value]);

                expect(props.updateTypeConfig).toHaveBeenCalledTimes(1);
                expect(props.updateTypeConfig).toHaveBeenCalledWith(
                    getTypeConfig(Types.set_contact_name)
                );
            });

            it('should update type config if passed an contact field asset', () => {
                const { wrapper, instance, props } = setup(true, { updateTypeConfig: setMock() });
                const existingField: Asset = {
                    id: 'field-0',
                    name: 'National ID',
                    type: AssetType.Field
                };

                instance.onChange([existingField]);

                expect(props.updateTypeConfig).toHaveBeenCalledTimes(1);
                expect(props.updateTypeConfig).toHaveBeenCalledWith(
                    getTypeConfig(Types.set_contact_field)
                );
            });

            it('should call onChange handler if passed', () => {
                const { wrapper, instance, props } = setup(true, { onChange: setMock() });

                instance.onChange([props.attribute]);

                expect(props.onChange).toHaveBeenCalledTimes(1);
                expect(props.onChange).toHaveBeenCalledWith(props.attribute);
            });

            it('should update typeConfig', () => {
                const setContactFieldAsset = fieldToAsset(createSetContactFieldAction().field);
                const { wrapper, instance, props } = setup(false, {
                    initial: { $set: setContactFieldAsset },
                    updateTypeConfig: setMock()
                });
                const setContactPropertyAsset = propertyToAsset(Types.set_contact_name);

                instance.onChange([setContactPropertyAsset]);

                expect(props.updateTypeConfig).toHaveBeenCalledTimes(1);
                expect(props.updateTypeConfig).toHaveBeenCalledWith(
                    getTypeConfig(Types.set_contact_name)
                );
            });
        });
    });
});
