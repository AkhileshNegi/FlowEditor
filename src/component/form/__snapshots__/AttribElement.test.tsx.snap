// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttribElement render should pass createOptions to SelectSearch 1`] = `
<FormElement
  attribError={false}
  errors={Array []}
  name="Attribute"
>
  <SelectSearch
    __className=""
    assets={
      FieldAssets {
        "assetType": "field",
        "assets": Object {
          "language": Object {
            "id": "language",
            "name": "Language",
            "type": "property",
          },
          "name": Object {
            "id": "name",
            "name": "Name",
            "type": "property",
          },
        },
        "endpoint": "/assets/fields.json",
        "idProperty": "key",
        "localStorage": true,
      }
    }
    closeOnSelect={true}
    createNewOption={[Function]}
    createPrompt="New attribute: "
    initial={
      Array [
        Object {
          "id": "name",
          "name": "Name",
          "type": "property",
        },
      ]
    }
    isOptionUnique={[Function]}
    isValidNewOption={[Function]}
    multi={false}
    name="Attribute"
    onChange={[Function]}
    placeholder="Enter the name of an existing attribute or create a new one"
    resultType="field"
    searchPromptText="Invalid attribute name"
  />
</FormElement>
`;

exports[`AttribElement render should render self, children with base props 1`] = `
Object {
  "__className": "",
  "assets": FieldAssets {
    "assetType": "field",
    "assets": Object {
      "language": Object {
        "id": "language",
        "name": "Language",
        "type": "property",
      },
      "name": Object {
        "id": "name",
        "name": "Name",
        "type": "property",
      },
    },
    "endpoint": "/assets/fields.json",
    "idProperty": "key",
    "localStorage": true,
  },
  "closeOnSelect": true,
  "initial": Array [
    Object {
      "id": "name",
      "name": "Name",
      "type": "property",
    },
  ],
  "multi": false,
  "name": "Attribute",
  "onChange": [Function],
  "placeholder": "Enter the name of an existing attribute or create a new one",
  "resultType": "field",
  "searchPromptText": "Invalid attribute name",
}
`;

exports[`AttribElement render should render self, children with base props 2`] = `
<FormElement
  attribError={false}
  errors={Array []}
  name="Attribute"
>
  <SelectSearch
    __className=""
    assets={
      FieldAssets {
        "assetType": "field",
        "assets": Object {
          "language": Object {
            "id": "language",
            "name": "Language",
            "type": "property",
          },
          "name": Object {
            "id": "name",
            "name": "Name",
            "type": "property",
          },
        },
        "endpoint": "/assets/fields.json",
        "idProperty": "key",
        "localStorage": true,
      }
    }
    closeOnSelect={true}
    initial={
      Array [
        Object {
          "id": "name",
          "name": "Name",
          "type": "property",
        },
      ]
    }
    multi={false}
    name="Attribute"
    onChange={[Function]}
    placeholder="Enter the name of an existing attribute or create a new one"
    resultType="field"
    searchPromptText="Invalid attribute name"
  />
</FormElement>
`;
